File: pipeline/src/fpl_dof/rules/legality.py
"""Squad legality, and the selling-price arithmetic.

:func:`validate_squad` returns **every** violation rather than the first. A squad with three
problems should tell you about three problems: the optimiser's job is to produce something legal in
one pass, and a validator that stops at the first failure makes debugging it a game of whack-a-mole.

This module is the thing everything else trusts. The optimiser's output is checked against it
rather than asserted to be correct by construction — a solver bug or a mis-stated constraint that
produces an illegal squad is exactly the class of error that would otherwise reach the deadline
unnoticed (DP-13).
"""

from __future__ import annotations

import math
from collections import Counter
from dataclasses import dataclass, field
from enum import StrEnum

from fpl_dof.rules.models import GameRules, Position, SquadRules


class ViolationCode(StrEnum):
    SQUAD_SIZE = "squad_size"
    DUPLICATE_PLAYER = "duplicate_player"
    COMPOSITION = "composition"
    BUDGET = "budget"
    CLUB_LIMIT = "club_limit"
    STARTING_SIZE = "starting_size"
    STARTER_NOT_IN_SQUAD = "starter_not_in_squad"
    FORMATION = "formation"
    CAPTAIN_NOT_STARTING = "captain_not_starting"
    VICE_NOT_STARTING = "vice_not_starting"
    CAPTAIN_IS_VICE = "captain_is_vice"
    BENCH_ORDER = "bench_order"


@dataclass(frozen=True, slots=True)
class Violation:
    code: ViolationCode
    message: str
    detail: dict[str, object] = field(default_factory=dict)


@dataclass(frozen=True, slots=True)
class SquadPlayer:
    player_id: int
    position: Position
    team_id: int
    price: float
    """The price at which this player counts against the budget."""


@dataclass(frozen=True, slots=True)
class Squad:
    """A full squad, plus how it is set up for a gameweek."""

    players: tuple[SquadPlayer, ...]
    starting: tuple[int, ...] = ()
    captain: int | None = None
    vice_captain: int | None = None
    bench_order: tuple[int, ...] = ()

    def by_id(self) -> dict[int, SquadPlayer]:
        return {player.player_id: player for player in self.players}

    @property
    def total_price(self) -> float:
        return round(sum(player.price for player in self.players), 1)

    def bench(self) -> tuple[int, ...]:
        starting = set(self.starting)
        return tuple(p.player_id for p in self.players if p.player_id not in starting)


def validate_squad(
    squad: Squad, rules: GameRules, *, budget: float | None = None
) -> list[Violation]:
    """Every way this squad breaks the rules, in a stable order.

    ``budget`` overrides the rules' initial budget — from E1 onward the relevant figure is squad
    value plus bank, not £100.0m.

    A squad with no starting XI set is checked for squad-level legality only. That is the E0 case:
    the optimiser picks 15 and a captain, and lineup checks apply once a lineup exists.
    """
    violations: list[Violation] = []
    squad_rules = rules.squad
    limit = squad_rules.budget if budget is None else budget

    violations.extend(_check_squad(squad, squad_rules, limit))
    if squad.starting:
        violations.extend(_check_lineup(squad, squad_rules))
    return violations


def _check_squad(squad: Squad, rules: SquadRules, budget: float) -> list[Violation]:
    violations: list[Violation] = []

    if len(squad.players) != rules.size:
        violations.append(
            Violation(
                ViolationCode.SQUAD_SIZE,
                f"squad has {len(squad.players)} players, expected {rules.size}",
                {"actual": len(squad.players), "expected": rules.size},
            )
        )

    ids = Counter(player.player_id for player in squad.players)
    duplicates = sorted(pid for pid, count in ids.items() if count > 1)
    if duplicates:
        violations.append(
            Violation(
                ViolationCode.DUPLICATE_PLAYER,
                f"player(s) selected more than once: {duplicates}",
                {"player_ids": duplicates},
            )
        )

    counts = Counter(player.position for player in squad.players)
    for position in Position:
        required = rules.composition[position]
        actual = counts.get(position, 0)
        if actual != required:
            violations.append(
                Violation(
                    ViolationCode.COMPOSITION,
                    f"{position.value}: {actual} selected, exactly {required} required",
                    {"position": position.value, "actual": actual, "expected": required},
                )
            )

    # Compare in tenths: 0.1 is not representable in binary, and a squad costing exactly the budget
    # must not be rejected because the sum came to 100.00000000000001.
    spend_tenths = sum(round(player.price * 10) for player in squad.players)
    budget_tenths = round(budget * 10)
    if spend_tenths > budget_tenths:
        violations.append(
            Violation(
                ViolationCode.BUDGET,
                f"squad costs £{spend_tenths / 10:.1f}m, budget is £{budget_tenths / 10:.1f}m",
                {"spend": spend_tenths / 10, "budget": budget_tenths / 10},
            )
        )

    per_club = Counter(player.team_id for player in squad.players)
    for team_id, count in sorted(per_club.items()):
        if count > rules.club_limit:
            violations.append(
                Violation(
                    ViolationCode.CLUB_LIMIT,
                    f"{count} players from club {team_id}, maximum is {rules.club_limit}",
                    {"team_id": team_id, "actual": count, "limit": rules.club_limit},
                )
            )

    return violations


def _check_lineup(squad: Squad, rules: SquadRules) -> list[Violation]:
    violations: list[Violation] = []
    members = squad.by_id()

    if len(squad.starting) != rules.starting_size:
        violations.append(
            Violation(
                ViolationCode.STARTING_SIZE,
                f"{len(squad.starting)} players starting, expected {rules.starting_size}",
                {"actual": len(squad.starting), "expected": rules.starting_size},
            )
        )

    outside = sorted(set(squad.starting) - set(members))
    if outside:
        violations.append(
            Violation(
                ViolationCode.STARTER_NOT_IN_SQUAD,
                f"starting player(s) not in the squad: {outside}",
                {"player_ids": outside},
            )
        )

    counts = Counter(members[pid].position for pid in squad.starting if pid in members)
    for position in Position:
        actual = counts.get(position, 0)
        low, high = rules.formation_min[position], rules.formation_max[position]
        if not low <= actual <= high:
            violations.append(
                Violation(
                    ViolationCode.FORMATION,
                    f"{position.value}: {actual} starting, must be between {low} and {high}",
                    {"position": position.value, "actual": actual, "min": low, "max": high},
                )
            )

    starting = set(squad.starting)
    if squad.captain is not None and squad.captain not in starting:
        violations.append(
            Violation(
                ViolationCode.CAPTAIN_NOT_STARTING,
                f"captain {squad.captain} is not in the starting XI",
                {"player_id": squad.captain},
            )
        )
    if squad.vice_captain is not None and squad.vice_captain not in starting:
        violations.append(
            Violation(
                ViolationCode.VICE_NOT_STARTING,
                f"vice-captain {squad.vice_captain} is not in the starting XI",
                {"player_id": squad.vice_captain},
            )
        )
    if (
        squad.captain is not None
        and squad.vice_captain is not None
        and squad.captain == squad.vice_captain
    ):
        violations.append(
            Violation(
                ViolationCode.CAPTAIN_IS_VICE,
                "captain and vice-captain are the same player",
                {"player_id": squad.captain},
            )
        )

    if squad.bench_order:
        expected = {
            pid
            for pid in squad.bench()
            if pid in members and members[pid].position is not Position.GKP
        }
        if set(squad.bench_order) != expected:
            violations.append(
                Violation(
                    ViolationCode.BENCH_ORDER,
                    "bench order must list exactly the outfield substitutes",
                    {"given": sorted(squad.bench_order), "expected": sorted(expected)},
                )
            )

    return violations


def is_legal(squad: Squad, rules: GameRules, *, budget: float | None = None) -> bool:
    return not validate_squad(squad, rules, budget=budget)


def selling_price(purchase_price: float, current_price: float, rules: SquadRules) -> float:
    """What a player sells for.

    Profit is shared with the game: you keep ``1 - sell_on_fee`` of any rise, rounded **down** to
    the nearest £0.1. A price fall is borne in full.

    Unused in E0 — a preseason squad has no purchase history — and needed from E1, which is why it
    is written and tested now rather than under deadline pressure later.
    """
    if rules.sell_at_purchase_price:
        return purchase_price

    purchase_tenths = round(purchase_price * 10)
    current_tenths = round(current_price * 10)
    if current_tenths <= purchase_tenths:
        return current_tenths / 10

    profit = current_tenths - purchase_tenths
    kept = math.floor(profit * (1.0 - rules.sell_on_fee))
    return (purchase_tenths + kept) / 10

File: web/src/components/squad/legality.ts
/**
 * Squad legality in the browser — the same rules the Python validator applies, read from the same
 * configuration (DL-14, Invariant 9, DP-05).
 *
 * **Nothing in this file knows what any FPL rule value is.** Squad size, starting size, budget, club
 * limit, per-position composition and the formation bounds all arrive as `rules.squad`, published
 * from the configuration `fpl_dof.rules` reads. A `3` for the club limit written here would be the
 * same defect as a `4` for a forward goal in a `.py` file, and would be right for exactly as long as
 * FPL left the rule alone.
 *
 * It is a deliberate mirror of `pipeline/src/fpl_dof/rules/legality.py`:
 *
 * - the **violation codes are the same strings** as that module's `ViolationCode`;
 * - the **`detail` keys are the same keys**, so the two can be compared structurally;
 * - every check **returns every violation** rather than stopping at the first, because a squad with
 *   three problems should say three things (a first-failure validator makes editing a game of
 *   whack-a-mole);
 * - the budget comparison is done **in tenths**, because 0.1 is not representable in binary and a
 *   squad costing exactly the budget must not be rejected for summing to 100.00000000000001.
 *
 * That alignment is the cheap half of the cross-language conformance test E6's definition of done
 * calls for: a harness need only feed both implementations the same squad and compare
 * `(code, detail)` pairs. Prose messages are deliberately *not* part of that contract — they are for
 * a reader, and the two languages phrase them for their own audiences.
 *
 * Pure by construction (DP-03): no I/O, no clock, no module state, configuration passed as an
 * argument.
 */

import type { Positionmap, Rules } from "../../contract/types";

export type Position = keyof Positionmap;

/**
 * The position codes, in the order a pitch reads.
 *
 * This is the contract's own key set — the four keys `Positionmap` declares — not a rule value. The
 * Python validator iterates its `Position` enum in exactly the same way. What each position
 * *requires* is read from the rules; only the identity of the positions is structural.
 */
export const POSITIONS: readonly Position[] = ["GKP", "DEF", "MID", "FWD"];

/** The same code set as `ViolationCode` in `fpl_dof.rules.legality`. */
export type ViolationCode =
  | "squad_size"
  | "duplicate_player"
  | "composition"
  | "budget"
  | "club_limit"
  | "starting_size"
  | "starter_not_in_squad"
  | "formation"
  | "captain_not_starting"
  | "vice_not_starting"
  | "captain_is_vice"
  | "bench_order";

export interface Violation {
  code: ViolationCode;
  /** For a reader. Not part of the cross-language contract. */
  message: string;
  /** For a machine. Keys match the Python validator's `detail` dictionary. */
  detail: Record<string, unknown>;
}

/** The minimum a squad member has to be, to be checkable. Everything else is presentation. */
export interface SquadMember {
  player_id: number;
  position: Position;
  team_id: number;
  /** The price at which this player counts against the budget. */
  price: number;
}

/** A squad, plus however much of a line-up has been decided so far. */
export interface ProposedSquad {
  players: readonly SquadMember[];
  /** Empty means "no line-up chosen yet" — squad-level checks only, as in the Python validator. */
  starting?: readonly number[];
  captain?: number | null;
  vice_captain?: number | null;
  bench_order?: readonly number[];
}

/** Prices summed in tenths, so a squad at exactly the budget is at exactly the budget. */
export function priceTenths(players: readonly SquadMember[]): number {
  return players.reduce((total, player) => total + Math.round(player.price * 10), 0);
}

export function totalPrice(players: readonly SquadMember[]): number {
  return priceTenths(players) / 10;
}

function countBy<T, K extends string | number>(items: readonly T[], key: (item: T) => K): Map<K, number> {
  const counts = new Map<K, number>();
  for (const item of items) {
    const k = key(item);
    counts.set(k, (counts.get(k) ?? 0) + 1);
  }
  return counts;
}

/**
 * Every way this squad breaks the rules, in a stable order.
 *
 * `budget` overrides the rules' initial budget: once a season is under way the relevant figure is
 * squad sell value plus bank, not the opening £100.0m, and the caller has that from `week.json`.
 *
 * A squad with no starting XI is checked for squad-level legality only.
 */
export function validateSquad(
  squad: ProposedSquad,
  rules: Rules,
  options: { budget?: number } = {},
): Violation[] {
  const squadRules = rules.squad;
  const budget = options.budget ?? squadRules.budget;
  const violations = checkSquad(squad.players, squadRules, budget);
  if (squad.starting && squad.starting.length > 0) {
    violations.push(...checkLineup(squad, squadRules));
  }
  return violations;
}

export function isLegal(squad: ProposedSquad, rules: Rules, options: { budget?: number } = {}): boolean {
  return validateSquad(squad, rules, options).length === 0;
}

function checkSquad(
  players: readonly SquadMember[],
  rules: Rules["squad"],
  budget: number,
): Violation[] {
  const violations: Violation[] = [];

  if (players.length !== rules.size) {
    violations.push({
      code: "squad_size",
      message: `squad has ${players.length} players, expected ${rules.size}`,
      detail: { actual: players.length, expected: rules.size },
    });
  }

  const ids = countBy(players, (player) => player.player_id);
  const duplicates = [...ids.entries()]
    .filter(([, count]) => count > 1)
    .map(([id]) => id)
    .sort((a, b) => a - b);
  if (duplicates.length > 0) {
    violations.push({
      code: "duplicate_player",
      message: `player(s) selected more than once: ${duplicates.join(", ")}`,
      detail: { player_ids: duplicates },
    });
  }

  const byPosition = countBy(players, (player) => player.position);
  for (const position of POSITIONS) {
    const required = rules.composition[position];
    const actual = byPosition.get(position) ?? 0;
    if (actual !== required) {
      violations.push({
        code: "composition",
        message: `${position}: ${actual} selected, exactly ${required} required`,
        detail: { position, actual, expected: required },
      });
    }
  }

  const spendTenths = priceTenths(players);
  const budgetTenths = Math.round(budget * 10);
  if (spendTenths > budgetTenths) {
    violations.push({
      code: "budget",
      message: `squad costs £${(spendTenths / 10).toFixed(1)}m, budget is £${(budgetTenths / 10).toFixed(1)}m`,
      detail: { spend: spendTenths / 10, budget: budgetTenths / 10 },
    });
  }

  const perClub = [...countBy(players, (player) => player.team_id).entries()].sort(
    (a, b) => a[0] - b[0],
  );
  for (const [teamId, count] of perClub) {
    if (count > rules.club_limit) {
      violations.push({
        code: "club_limit",
        message: `${count} players from club ${teamId}, maximum is ${rules.club_limit}`,
        detail: { team_id: teamId, actual: count, limit: rules.club_limit },
      });
    }
  }

  return violations;
}

function checkLineup(squad: ProposedSquad, rules: Rules["squad"]): Violation[] {
  const violations: Violation[] = [];
  const members = new Map(squad.players.map((player) => [player.player_id, player]));
  const starting = squad.starting ?? [];

  if (starting.length !== rules.starting_size) {
    violations.push({
      code: "starting_size",
      message: `${starting.length} players starting, expected ${rules.starting_size}`,
      detail: { actual: starting.length, expected: rules.starting_size },
    });
  }

  const outside = [...new Set(starting.filter((id) => !members.has(id)))].sort((a, b) => a - b);
  if (outside.length > 0) {
    violations.push({
      code: "starter_not_in_squad",
      message: `starting player(s) not in the squad: ${outside.join(", ")}`,
      detail: { player_ids: outside },
    });
  }

  const startingPositions = countBy(
    starting.filter((id) => members.has(id)),
    (id) => members.get(id)!.position,
  );
  for (const position of POSITIONS) {
    const actual = startingPositions.get(position) ?? 0;
    const low = rules.formation_min[position];
    const high = rules.formation_max[position];
    if (actual < low || actual > high) {
      violations.push({
        code: "formation",
        message: `${position}: ${actual} starting, must be between ${low} and ${high}`,
        detail: { position, actual, min: low, max: high },
      });
    }
  }

  const startingSet = new Set(starting);
  const captain = squad.captain ?? null;
  const vice = squad.vice_captain ?? null;

  if (captain !== null && !startingSet.has(captain)) {
    violations.push({
      code: "captain_not_starting",
      message: `captain ${captain} is not in the starting XI`,
      detail: { player_id: captain },
    });
  }
  if (vice !== null && !startingSet.has(vice)) {
    violations.push({
      code: "vice_not_starting",
      message: `vice-captain ${vice} is not in the starting XI`,
      detail: { player_id: vice },
    });
  }
  if (captain !== null && vice !== null && captain === vice) {
    violations.push({
      code: "captain_is_vice",
      message: "captain and vice-captain are the same player",
      detail: { player_id: captain },
    });
  }

  if (squad.bench_order && squad.bench_order.length > 0) {
    // Compared as *sets*, which is what the Python validator does — a deliberate match, not a
    // convenience. Comparing sorted lists instead reports a violation for a bench order that names
    // one substitute twice, and the Python authority does not; the cross-language corpus caught the
    // disagreement and this is the side that changed (Invariant 9, DL-39). Both are therefore lax
    // about a repeated id, which is recorded rather than hidden: nothing can produce one, because
    // `draft.ts` derives the bench order from the squad rather than accepting one from a caller.
    const expectedSet = new Set(
      squad.players
        .filter((player) => !startingSet.has(player.player_id) && player.position !== "GKP")
        .map((player) => player.player_id),
    );
    const givenSet = new Set(squad.bench_order);
    const expected = [...expectedSet].sort((a, b) => a - b);
    const given = [...squad.bench_order].sort((a, b) => a - b);
    const same =
      givenSet.size === expectedSet.size && [...givenSet].every((id) => expectedSet.has(id));
    if (!same) {
      violations.push({
        code: "bench_order",
        message: "bench order must list exactly the outfield substitutes",
        detail: { given, expected },
      });
    }
  }

  return violations;
}

/**
 * Every starting XI shape the rules permit, in a stable order.
 *
 * Mirrors `SquadRules.legal_formations`: the goalkeeper, defender and midfielder counts are
 * enumerated within their published bounds and the forward count is whatever is left of the
 * starting size, kept only when it too falls within its bounds. Nothing here assumes 1-3-2-1 or
 * eleven starters — change the configuration and this changes with it.
 */
export function legalFormations(rules: Rules): Array<Record<Position, number>> {
  const squad = rules.squad;
  const formations: Array<Record<Position, number>> = [];
  for (let gkp = squad.formation_min.GKP; gkp <= squad.formation_max.GKP; gkp += 1) {
    for (let def = squad.formation_min.DEF; def <= squad.formation_max.DEF; def += 1) {
      for (let mid = squad.formation_min.MID; mid <= squad.formation_max.MID; mid += 1) {
        const fwd = squad.starting_size - gkp - def - mid;
        if (fwd >= squad.formation_min.FWD && fwd <= squad.formation_max.FWD) {
          formations.push({ GKP: gkp, DEF: def, MID: mid, FWD: fwd });
        }
      }
    }
  }
  return formations;
}

/**
 * What a player sells for — the mirror of `selling_price` in the Python rules module.
 *
 * Profit is shared with the game: the manager keeps `1 - sell_on_fee` of any rise, rounded **down**
 * to the nearest £0.1; a fall is borne in full. Both the fee and whether selling is at purchase
 * price at all come from the rules.
 *
 * Present and tested here, but **not yet applied by the builder**: the published contract carries no
 * per-player purchase price, so the browser cannot reconstruct what any individual player would
 * raise. `week.squad_state.budget` already nets the fee off in aggregate, which is what the builder
 * spends against; see the note in `SquadBuilder`. When a purchase price reaches the contract, this
 * is the function that consumes it.
 */
export function sellingPrice(purchasePrice: number, currentPrice: number, rules: Rules): number {
  if (rules.squad.sell_at_purchase_price) return purchasePrice;

  const purchaseTenths = Math.round(purchasePrice * 10);
  const currentTenths = Math.round(currentPrice * 10);
  if (currentTenths <= purchaseTenths) return currentTenths / 10;

  const profit = currentTenths - purchaseTenths;
  const kept = Math.floor(profit * (1 - rules.squad.sell_on_fee));
  return (purchaseTenths + kept) / 10;
}
